### ISSUE TITLE
Headless Client for Smart Speakers

### ISSUE THREAD

**Issue by MtotheKay** [SEPIA-Framework/sepia-docs#9]
Hi Florian,

first of all: amazing work. I was searching for a stt engine that can use other languages than English to use for my smart home, and stumbled via Zambia onto your master piece.

I've read on the home page that SEPIA could be used for smart speakers, but could not find a headless client version on github. Are you planning for something like that? 

Best Regards
Martijn

**Comment by MtotheKay** [comment]
Hi Florian,

sounds great. As long as it works on a pi 3, I would use it.

I guess the event handling for a headless client needs to be handled differently then for a HTML Client e.G. Lists need to be spoken instead of shown... or which might also be a possibility -> show the List on a different device.

**Comment by fquirin** [comment]
The basic idea is that each service can check if the client has a display (there is a Java function in the ServiceBuilder somewhere) and adjust the answer accordingly, e.g. don't show any result cards at all etc. ... though I like the idea of "show[ing] the List on a different device" (this should be doable via the chat-server somehow ^^) :+1: 

I'll do some testing on RPi3 and RPi4 and will update the issue with some more info soon :-)

**Comment by ocdtrekkie** [comment]
I'm super interested in this. I bought a bunch of Google AIY kits that were $3 a pop at my local Micro Center, just for the voice hat, microphone, fancy box, etc. and I'd love to turn them into internal network only voice clients.

In my case, I have a weird one though: I wrote the assistant I'd like to do all of the "logic" already. Handling voice is out of my league though. The fact SEPIA clients offer the command output over a websocket after doing all of the wake word and STT makes it a particularly interesting front end to potentially connect to other backends as well.

**Comment by fquirin** [comment]
I'd be very interested to learn how well the Google AIY kits work with SEPIA :-) Maybe I can help to put everything together.

The "logic" you've mentioned, does that include the NLU and service integration? What programming language do you use?

SEPIA has a module for intent classification + NER, a dialog manager and a service execution module. Intent + NER is basically a chain of interpretation steps and currently I'm planning to add an additional step for external API calls (which could be used to hook-up your logic if it does NLU). If you want to run your own service as well this could maybe done with kind of a dummy service that just fetches the result from your program.

**Comment by ocdtrekkie** [comment]
If you live near at Micro Center, the version 1 kits are $3: https://www.microcenter.com/product/483414/aiy-voice-kit

I've seen a few other assistant platforms tried with them, so it should be possible. You bring your own Raspberry Pi, but it's got the audio chips and a dual microphone array and the button and the box. I've seen a Snips tutorial, and Mycroft has a setup option for using AIY kits built-in.

I'm not doing any heavy NLU, since my app is currently for personal use, I'm happy with somewhat rigid commands, though I may play with an NLP tagger and stuff in the future. I have a Visual Basic .NET app that runs off text input in plain English, and currently can also accept that input from a command line argument or an email.

So I figure I can either implement a way to take input from another source, or if doing that in VB is challenging, I can always have a little piece of code in something else that's an intermediate between the two. Current I do already do TTS output, though obviously I'd want to feed that text back to any clients so the speech response came from the same client that received the request.

**Comment by fquirin** [comment]
> If you live near at Micro Center, the version 1 kits are $3: https://www.microcenter.com/product/483414/aiy-voice-kit

Unfortunately we don't have those in Germany ^^ but I'll keep the eyes open. I've seen a few now and then online for ~15€ I think.

> So I figure I can either implement a way to take input from another source, or if doing that in VB is challenging, I can always have a little piece of code in something else that's an intermediate between the two. 

In theory it should be possible for a [SEPIA Mesh-Node](https://github.com/SEPIA-Framework/sepia-mesh-nodes) to call your command-line interface via the `RuntimePlugin`. I wrote about it in a slightly different context [here](https://medium.com/sepia-framework/control-your-windows-system-remotely-via-voice-a1cc40b794).
It might take a bit of experimenting but in theory the idea is:
* The Mesh-Node runs next to your program with access to the command-line interface
* It creates a server that can be safely accessed via the SEPIA Assist-Server
* You define a command via Teach-UI or a custom service via the SEPIA SDK that calls the Mesh-Node
* You get your answer right inside the app :-)

**Comment by fquirin** [comment]
>  I've decided to focus on the Chromium-Kiosk-Mode version for RaspberryPi first. In theory this option can work even without any display connected (with a dummy X-server like Xvfb) but obviously eats up a few MB more RAM. The advantage is that all the microphone handling, STT, TTS and web features will keep on working. There is a client info variable as well that can be used by services to check if the client has support for a display to tweak the answer results which I'm planning to expose as an URL parameter.

I was pretty busy with the latest SEPIA release but I finally got some time now to play around with this idea and actually it works pretty well :-) First I thought it could work with Chromium '--headless' option (which is pretty cool) ... and it kind of does ... until it doesn't: no support for TTS :-( ... but Xvfb is super easy to install and run:
```
sudo apt-get install xvfb
xvfb-run chromium-browser --disable-features=VizDisplayCompositor --no-sandbox --autoplay-policy=no-user-gesture-required --user-data-dir='/home/pi/cache/chromium_sepia' --enable-speech-dispatcher --kiosk 'https://sepia-framework.github.io/app/launcher.html?autolaunch=true'
```

`--disable-features=VizDisplayCompositor` was required for the latest Raspbian and Chromium but might not be needed in older versions. `--no-sandbox` might also be unnecessary, I need to test this more.
The whole procedure currently looks like this:
* Connect your RPi to a monitor
* Start Chromium, login to SEPIA and setup everything as you like it (be sure to grant microphone access etc.)
* Close Chromium, put the xvfb-run command in autostart and disconnect monitor
* Reboot

I have an unused ReSpeaker 2-Mics Pi HAT flying around an will try to put together a complete tutorial soon testing everything thoroughly :-)

**Comment by sepia-assistant** [comment]
The beta version for the headless RPi client is now available :-)
https://github.com/SEPIA-Framework/sepia-installation-and-setup/tree/master/sepia-client-installation/rpi

**Comment by eX00r** [comment]
Wonderful! just one probably noob question: Is it possible to install this client alongside the Sepia server on the same raspi 4? 

Right now I only have one device and was wondering if I will be able to form it into a fully functional and listening smart assistant.

**Comment by sepia-assistant** [comment]
Yes that's possible if you have the 2GB or 4GB Version of the Pi4 :slightly_smiling_face:

**Comment by Smarthome-Creator** [comment]
Hello,
I have a little problem with DIY. Unfortunately, the DIY ignored me again and again so I looked at the log in the clexi and found something.
If the TV or music runs on the side and the DIY triggers by itself then it gets stuck in the "listening" status.
All I can do is send a "call test" to put it back into the state active state.
The same happens in the status "type": "tts_speak", "msg".
If there are no background noises then it runs smoothly.

Do you have any ideas how I can get this under control?

**Comment by sepia-assistant** [comment]
I've heard of a similar problem in the FHEM forum. For the new version I've implemented a few auto-recovery procedures ... maybe this will fix it or at least repair it after a few seconds when it happens.

> The same happens in the status "type": "tts_speak", "msg".

This is really strange. Is there any other error message? I've never seen the client get stuck in "speak" mode :see_no_evil:

**Comment by Smarthome-Creator** [comment]
Sorry i copied the wrong line. Here is the right on:

{"type":"asr_result","msg":""}}
